### ISSUE TITLE
nosto_product_sync index is stuck.

### ISSUE THREAD

**Issue by pauloharaujos** [Nosto/nosto-magento2#507]
**Magento 2 version(s) used**: 2.3.2
**Extension version(s) affected**: 3.7.5

**Description**  
When I try to edit and save a product in admin, it takes a long time, often over 10 minutes. 

![loading](https://user-images.githubusercontent.com/23405617/61898748-65835280-aef0-11e9-8621-9eb9c3a4b301.png)

But if I change Nosto indexer mode for Update By Schedule, the error disappears. Also, When I try to run this command: `bin/magento index: reindex nosto_product_sync`, the nosto product update index gets stuck in "Processing".

![nostoIndex](https://user-images.githubusercontent.com/23405617/61898752-674d1600-aef0-11e9-8b09-405be529ec6e.png)

**How to reproduce**  
1- Change Nosto indexer mode to Update on Save.
2- Log on the Magento Admin Panel. Go to Catalog -> Products
3-  Open a product, edit description and click on the save button.

**Magento 2 mode**
- [X] Developer 
- [X] Production

**Full page cache**
- [X] Enabled 
- [ ] Disabled

**Nosto indexer mode**
- [X] Update On Save  
- [ ] Update By Schedule

**Comment by hannunosto** [comment]
Hi @pauloharaujos, can you try to truncate tables `nosto_tagging_product_queue` and `nosto_product_sync_cl`. These tables might get crowded when switching between "Update on save" and "Update by schedule". We are working on a complete indexer overhaul that should fix these issues

**Comment by andreymoskvenkov** [comment]
:-D
Yes, the nosto indexer sucks.
Looks like this team completely do not understand how to build good indexers. 
NEVER!!! add triggers on Magento index tables, because each time Magento reindexes records, 1 000 000 of IDs fall into changelog table + causes deadlocks during reindexing of price and catalogrule indexers.
Use custom plugin on execute() method and collect Really changed product IDs via making a snapshot of prce index Before reindex and then compare it with the final table After reindex (of price index) and add really changed IDs to Nosto changelog table at the end. 
This is what we do in all our indexers implementation, when we need to refect changes in other indexers.

**Comment by hannunosto** [comment]
> :-D
> Yes, the nosto indexer sucks.
> Looks like this team completely does not understand how to build good indexers.
> NEVER!!! add triggers on Magento index tables, because each time Magento reindexes records, 1 000 000 of IDs fall into changelog table + causes deadlocks during reindexing of price and catalogrule indexers.
> Use custom plugin on execute() method and collect Really changed product IDs via making a snapshot of prce index Before reindex and then compare it with the final table After reindex (of price index) and add really changed IDs to Nosto changelog table at the end.
> This is what we do in all our indexers implementation, when we need to refect changes in other indexers.

Thank you for your comments @andreymoskvenkov. We are working on an indexing logic rewrite at the moment (to be released in 4.0.0) but all feedback and improvement ideas are more than welcome. We're not sure tough what you mean by "Use custom plugin on execute() method and collect Really changed product IDs via making a snapshot of prce index". Could you please link here some of your indexer implementations so we could have a look if fits our use case as well? That would be helpful.

**Comment by hannunosto** [comment]
> Also "load" of URL rewrites, prices, stock items, categories one-by-one for EACH product and each store makes the indexer working 2 days on 100 000 products and 5 stores on our system. Need to load all needed information in the product collection (like: ->addUrlRewrites() or ->addCategoryIds()) or via direct queries by the list of product IDs. Otherwise each product takes 200 ms, which is: 100 000 products x 0.2 x 5 stores = 100 000 seconds = 2 days.
> Sorry, but I cannot share that code. You have to think by your own brain. I can share only the idea: when Price index starts reindexing, need to save all rows from catalog_product_price_index table into some temporary table and then at the end of price reindex need to select new records for the same product IDs from the updated catalog_product_price_index table and then compare prices and if a price for the same product_id, website_id and customer_group_id was changed, then insert this product ID into nosto_cl table. This logic is kind of moving triggers from index tables into PHP.
> I have already optimized your indexer to 1 hour for now (for now I haven't yet rewritten load of url rewrites and categories from one-by-one into a single query), but at least stock items were pre-warmed and loading products from the repository was removed. 3.8.8.
> But I cannot share it, sorry. NDA. I can help only by share the idea and I have already done it. Next step is your's.

Ok, thanks for you ideas @andreymoskvenkov. We'll look into this.

**Comment by IvanChepurnyi** [comment]
What's the purpose of the index in the first place? I remember it was trying to save data to the index to speed up lookup on frontend by after fixes suggested by me in #593 I am not sure you really need this index.

**Comment by IvanChepurnyi** [comment]
@andreymoskvenkov I think you are too rude and unprofessional. I understand you are frustrated with their indexer's performance, but neither it or NDA gives you the privilege to bash other developers. Not every developer supposed to know how to write efficient indexes, politely explain, and professionally share your point of view. We all are human beings.

**Comment by imgali** [comment]
i agree with @IvanChepurnyi sentiment. All the folks at Nosto have been open and professional. They have been working really hard to solve many of these problems whereas many extension vendors would never even respond or make an attempt. Its a pleasure working with the team at Nosto.

**Comment by andreymoskvenkov** [comment]
@IvanChepurnyi You are completely right about me.
I can confirm it by the fact, that I completed the indexer's optimization in 4 days, as it touches so many areas and the result is: 50 minutes of full reindex on Production, 4 stores and 100 000 products.
I will continue doing unprofessional things.

**Comment by IvanChepurnyi** [comment]
@andreymoskvenkov that's great that you did this improvement for your custom project. I also did a lot of high-performance work for various customers, but that does not give me the privilege to be rude to others. Please keep this conversation professional.

**Comment by andreymoskvenkov** [comment]
My life - my rules, sorry.

**Comment by mridang** [comment]
@IvanChepurnyi @andreymoskvenkov @imgali We've made a new release with the changes suggested by @IvanChepurnyi and this now available as 3.10.0 and also the new parallel-indexer in 4.0.0-RC4. We're still running performance tests internally to see how much we've improved.
In the next version, we will try to tackle https://github.com/Nosto/nosto-magento2/issues/308 
Your feedback and contribution is very welcome. 🙏

**Comment by imgali** [comment]
@mridang that is fantastic news. I will have our team start to work with it in a staging environment and try to provide some feedback. Thank you for all your hard work with getting this resolved.

**Comment by hannunosto** [comment]
Thanks for commenting on this one. The old indexer is replaced with two separate indexer in our next major release (currently https://github.com/Nosto/nosto-magento2/releases/tag/4.0.0-rc6). We will update our wiki docs during the actual release in more detail but basically the we've improved constructing the data needed for Nosto product for category lookups, sku lookups, price lookups (for both SKUs and "main product") and attribute lookups. 

As @mridang mentioned above both new indexers can be run in parallel mode. In the latest RC (4.0.0-rc6) it's also possible to set the somewhat heavy operation of constructing product data to be done in a cron job rather than in an indexer.     

Please comment here, if you feel we should reopen this issue.
